modules/standard/ChapelVersion.chpl

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright 2020 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+
+This module contains features that support reasoning about Chapel
+version numbers at compile-time.  Chapel version numbers have major,
+minor, and update components as in semantic versioning, though for
+Chapel, version 2.0.0 corresponds to version 1.0.0 in traditional
+semantic versioning.  In addition, pre-release versions of Chapel
+(those built from the GitHub source tree) have a SHA component
+indicating a specific point in that version's development.
+
+This module features:
+
+* :var:`chplVersion`: the version number of the copy of ``chpl`` used
+  to compile the program.
+
+* :proc:`version`: a utility function for creating new version values
+
+* :type:`versionValue`: the record type used to represent version
+  values, supporting compile-time comparisons
+
+*/
+
+module ChapelVersion {
+  // Query the compiler to get its version number and cache those values
+  private param major = __primitive("version major");
+  private param minor = __primitive("version minor");
+  private param update = __primitive("version update");
+  private param SHA = __primitive("version sha");
+
+
+  /* The version of ``chpl`` used to compile this program. */
+  const chplVersion: versionValue(?);
+  chplVersion = new versionValue(major, minor, update, SHA);
+
+  
+  /*
+    A helper function that creates a new version value from its
+    arguments.
+
+    :arg major: The major version number
+    :type major: `int`
+
+    :arg minor: The minor version number
+    :type minor: `int`
+
+    :arg update: The optional update version number (defaults to 0)
+    :type major: `int`
+
+    :arg sha: The optional SHA (defaults to "")
+    :type major: `int`
+
+    :returns: A new version value
+  */
+     
+  proc version(param major: int, param minor: int,
```

Comment:
Any reason to not have `minor` default to 0 as well?

Reply:
There's no real reason, no.  I was thinking in practice any reasonable query of a version number would want to compare at least major and minor and that `version(2,0)` said that more clearly than `version(2)`.  I worry a little that `version(2)` might suggest to someone `2.*.*` (any version in the 2.x.y family) rather than `2.0.0`.

Reply:
I've captured this question in the OP as a design question to converge on.